### DISCUSSION TITLE
Media upload with remote sidekiq workers

### DISCUSSION THREAD

**Discussion by leahoswald** [mastodon/mastodon#19500] [Q&A] [answered]
If I have an instance with remote sidekiq workers do they need a shared upload directory or service? Which queue is used for media processing? And is it possible to create a queue specific for media processing?

**Comment by Gargron** [comment, marked as answer]
To use Sidekiq on a separate machine than web workers, you need to store files in object storage (S3 or similar). There might be a way to create a shared filesystem some other way but it sounds more complicated. The default queue is used for media processing.

**Reply by leahoswald** [reply]
Well would be great if there was a separate queue for that.

**Reply by nigini** [reply]
I got to the same conclusion (except that I am not knowledgeable of Sidekq and Ruby.) But considering that media processing is getting in the way of other synchronous calls, what would be the easiest way to scale things?

- Can Sidekiq run a queue in other machines? (That is likely the easiest/cheapest way to scale?)
- Should one focus on deploying smaller servers behind a Load Balancer (which may be expensive)
- What are others doing?

**Reply by leahoswald** [reply]
We currently use on external worker server that is connected via wireguard becuase the redis lib used by mastodon can't use TLS. We will add another in the next few days. We also moved the DB to another server and we have installed Minio for S3 Storage. I plan to write a an article on scaling mastodon as soon as I find the time.

**Reply by supernovae** [reply]
> We currently use on external worker server that is connected via wireguard becuase the redis lib used by mastodon can't use TLS. We will add another in the next few days. We also moved the DB to another server and we have installed Minio for S3 Storage. I plan to write a an article on scaling mastodon as soon as I find the time.

Fantastic! look forward to hearing how you solved some of the scaling challenges.

**Reply by tuetenk0pp** [reply]
I know its a bit dated but theres an article about scaling on the official blog: https://blog.joinmastodon.org/2017/04/scaling-mastodon/

**Reply by supernovae** [reply]
> I know its a bit dated but theres an article about scaling on the official blog: https://blog.joinmastodon.org/2017/04/scaling-mastodon/

Yeah, i saw that.  Is using the local s3 separate from defining s3 for media files? and does anyone have any nginx config for how to split streaming and web workers? i'm assumeing its all endpoints and a load balancer can handle that (need to go look in code - but just figured i'd ask)

**Comment by supernovae** [comment]
If i use minio to distributed between clusters, can that be separate from the media s3?  i'd like them to be separate so local cache/content could be on a local distribution and also reduce my penalty for deletion early on a commercial s3 service.
